# Novo Nordisk Inc.

<p align="left">
  <img width="800" height="400" src="https://www.luiss.it/sites/www.luiss.it/files/images/notices/30/NOVO.png"> 
</p>

### Ankur Patel ;)

In [187]:
import pandas as pd

import warnings
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', 999)

### IMPORT VOLUME CALCULATIONS

In [188]:
vol = pd.read_excel("QAP RE19_1 Volume Calcs XPO Restated.xlsx", sheet_name="YTD_MTD for Indexes")
vol

Current Month           Year Start     Unnamed: 2    Unnamed: 3  \
0   2019-04-01 00:00:00  2019-01-01 00:00:00            NaN           NaN   
1                   NaN                  NaN            NaN           NaN   
2                   NaN         Year-to-Date            NaN           NaN   
3                   NaN       Demand\nVolume  Volume\nShare  Comm. Volume   
4              Victoza®          1.74911e+06       0.340187        997775   
5                   NaN                  NaN            NaN           NaN   
6              Ozempic®               503644      0.0979543        381025   
7                   NaN                  NaN            NaN           NaN   
8              Tresiba®              2713.68       0.135247       1558.09   
9                   NaN                  NaN            NaN           NaN   
10             Saxenda®               138824        0.28136        132975   
11                  NaN                  NaN            NaN           NaN   
12            Xultophy®               77.917     0.00388331       56.5375   
13                  NaN                  NaN            NaN           NaN   
14             Levemir®              3939.66       0.196349       1749.67   
15                  NaN                  NaN            NaN           NaN   
16             NovoLog®              5973.53        0.41628       3344.92   
17                  NaN                  NaN            NaN           NaN   
18               Fiasp®               123.52     0.00860778       105.412   
19                  NaN                  NaN            NaN           NaN   
20         NovoLog® Mix              927.632        0.29095       360.741   
21                  NaN                  NaN            NaN           NaN   
22                  NaN                  NaN            NaN           NaN   

       Unnamed: 4  Unnamed: 5      Unnamed: 6     Unnamed: 7    Unnamed: 8  \
0             NaN         NaN             NaN            NaN           NaN   
1             NaN         NaN             NaN            NaN           NaN   
2             NaN         NaN   Month-to-Date            NaN           NaN   
3   Part D Volume         NaN  Demand\nVolume  Volume\nShare  Comm. Volume   
4          580523         NaN          424969       0.326928        241353   
5             NaN         NaN             NaN            NaN           NaN   
6         95766.7         NaN          141872       0.109142        106870   
7             NaN         NaN             NaN            NaN           NaN   
8         999.547         NaN         695.553       0.140641       398.218   
9             NaN         NaN             NaN            NaN           NaN   
10        2616.07         NaN         36984.7       0.295034       35464.5   
11            NaN         NaN             NaN            NaN           NaN   
12        18.7773         NaN         20.1529     0.00407491       14.5616   
13            NaN         NaN             NaN            NaN           NaN   
14        1855.49         NaN         970.526        0.19624        433.33   
15            NaN         NaN             NaN            NaN           NaN   
16        2038.92         NaN         1481.01       0.415494       840.651   
17            NaN         NaN             NaN            NaN           NaN   
18        14.6972         NaN         33.6748     0.00944742       28.7842   
19            NaN         NaN             NaN            NaN           NaN   
20        463.869         NaN         228.272       0.292217       88.6443   
21            NaN         NaN             NaN            NaN           NaN   
22            NaN         NaN             NaN            NaN           NaN   

       Unnamed: 9  Unnamed: 10  Unnamed: 11  Unnamed: 12  Unnamed: 13  \
0             NaN          NaN          NaN          NaN          NaN   
1             NaN          NaN          NaN          NaN          NaN   
2             NaN          NaN          NaN          NaN          NaN

In [189]:
vol.dropna(how="all")

Current Month           Year Start     Unnamed: 2    Unnamed: 3  \
0   2019-04-01 00:00:00  2019-01-01 00:00:00            NaN           NaN   
1                   NaN                  NaN            NaN           NaN   
2                   NaN         Year-to-Date            NaN           NaN   
3                   NaN       Demand\nVolume  Volume\nShare  Comm. Volume   
4              Victoza®          1.74911e+06       0.340187        997775   
5                   NaN                  NaN            NaN           NaN   
6              Ozempic®               503644      0.0979543        381025   
7                   NaN                  NaN            NaN           NaN   
8              Tresiba®              2713.68       0.135247       1558.09   
9                   NaN                  NaN            NaN           NaN   
10             Saxenda®               138824        0.28136        132975   
11                  NaN                  NaN            NaN           NaN   
12            Xultophy®               77.917     0.00388331       56.5375   
13                  NaN                  NaN            NaN           NaN   
14             Levemir®              3939.66       0.196349       1749.67   
15                  NaN                  NaN            NaN           NaN   
16             NovoLog®              5973.53        0.41628       3344.92   
17                  NaN                  NaN            NaN           NaN   
18               Fiasp®               123.52     0.00860778       105.412   
19                  NaN                  NaN            NaN           NaN   
20         NovoLog® Mix              927.632        0.29095       360.741   
21                  NaN                  NaN            NaN           NaN   
22                  NaN                  NaN            NaN           NaN   

       Unnamed: 4  Unnamed: 5      Unnamed: 6     Unnamed: 7    Unnamed: 8  \
0             NaN         NaN             NaN            NaN           NaN   
1             NaN         NaN             NaN            NaN           NaN   
2             NaN         NaN   Month-to-Date            NaN           NaN   
3   Part D Volume         NaN  Demand\nVolume  Volume\nShare  Comm. Volume   
4          580523         NaN          424969       0.326928        241353   
5             NaN         NaN             NaN            NaN           NaN   
6         95766.7         NaN          141872       0.109142        106870   
7             NaN         NaN             NaN            NaN           NaN   
8         999.547         NaN         695.553       0.140641       398.218   
9             NaN         NaN             NaN            NaN           NaN   
10        2616.07         NaN         36984.7       0.295034       35464.5   
11            NaN         NaN             NaN            NaN           NaN   
12        18.7773         NaN         20.1529     0.00407491       14.5616   
13            NaN         NaN             NaN            NaN           NaN   
14        1855.49         NaN         970.526        0.19624        433.33   
15            NaN         NaN             NaN            NaN           NaN   
16        2038.92         NaN         1481.01       0.415494       840.651   
17            NaN         NaN             NaN            NaN           NaN   
18        14.6972         NaN         33.6748     0.00944742       28.7842   
19            NaN         NaN             NaN            NaN           NaN   
20        463.869         NaN         228.272       0.292217       88.6443   
21            NaN         NaN             NaN            NaN           NaN   
22            NaN         NaN             NaN            NaN           NaN   

       Unnamed: 9  Unnamed: 10  Unnamed: 11  Unnamed: 12  Unnamed: 13  \
0             NaN          NaN          NaN          NaN          NaN   
1             NaN          NaN          NaN          NaN          NaN   
2             NaN          NaN          NaN          NaN          NaN

In [190]:
vol["Current Month"][vol["Current Month"].notnull()]

0     2019-04-01 00:00:00
4                Victoza®
6                Ozempic®
8                Tresiba®
10               Saxenda®
12              Xultophy®
14               Levemir®
16               NovoLog®
18                 Fiasp®
20           NovoLog® Mix
Name: Current Month, dtype: object

### GET TRESIBA VOLUME

In [191]:
tr = vol.iloc[8,1:10]
tr

Year Start     2713.68
Unnamed: 2    0.135247
Unnamed: 3     1558.09
Unnamed: 4     999.547
Unnamed: 5         NaN
Unnamed: 6     695.553
Unnamed: 7    0.140641
Unnamed: 8     398.218
Unnamed: 9     257.753
Name: 8, dtype: object

In [192]:
for i in tr:
    print(i)

2713.682190686627
0.1352474942410895
1558.0904349881093
999.5472062234786
nan
695.5531632436376
0.14064069675271818
398.2175630202265
257.7528579685685


In [193]:
tres = [tr[0],"","",tr[1],"","",tr[2],"","",tr[3]]
tres

[2713.682190686627,
 '',
 '',
 0.1352474942410895,
 '',
 '',
 1558.0904349881093,
 '',
 '',
 999.5472062234786]

### IMPORT PERFORMANCE FILE TO UPDATE

In [194]:
may19_3D = pd.read_excel("QAP May 2019 NNI Performance Business Update v2.xlsm", sheet_name="3D. Basal in-market")
may19_3D

Unnamed: 0  \
0   Tresiba® continues strong volume and share per...   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
5                                                 NaN   
6                                                 NaN   
7                                                 NaN   
8                                                 NaN   
9                                                 NaN   
10                                                NaN   
11                                                NaN   
12                                                NaN   
13                                                NaN   
14                                                NaN   
15                                                NaN   
16                                                NaN   
17                                                NaN   
18                                                NaN   
19                                                NaN   
20                                                NaN   
21                                                NaN   
22                                                NaN   
23                                                NaN   
24                                                NaN   
25                                                NaN   
26                                                NaN   
27                                                NaN   
28                                                NaN   
29                                                NaN   
30                                                NaN   
31                                                NaN   
32                                                NaN   
33                                                NaN   
34                                                NaN   
35                                                NaN   
36                                                NaN   
37                                                NaN   
38                                                NaN   

                                           Unnamed: 1 Unnamed: 2 Unnamed: 3  \
0                                                 NaN        NaN        NaN   
1          Tresiba® demand volume overview (YTD May*)        NaN        NaN   
2                                                 NaN        NaN        NaN   
3                        Tresiba® demand vol.\n(mXPO)   Realized     RE19/1   
4                                                 NaN        NaN        NaN   
5                                        Total volume    2720.97    2713.68   
6                                                 NaN        NaN        NaN   
7                                                 NaN        NaN        NaN   
8                        Total share of basal segment   0.135579   0.135247   
9                                                 NaN        NaN        NaN   
10                                                NaN        NaN        NaN   
11                                  Commercial volume    1557.52    1558.09   
12                                                NaN        NaN        NaN   
13                                                NaN        NaN        NaN   
14                                      Part D volume    1006.14    999.547   
15                                                NaN        NaN        NaN   
16                                                NaN        NaN        NaN   
17                                                NaN        NaN        NaN   
18                                                NaN        NaN        NaN   
19    *March 2019 represents the most current mont...        NaN        NaN   
20                                                NaN        NaN        NaN   
21  Retail: Tresiba® sou

### UPDATE TRESIBA VOLUME
### 3D. Basal in-market

In [195]:
may19_3D.iloc[5:15,3] = tres
may19_3D

Unnamed: 0  \
0   Tresiba® continues strong volume and share per...   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
5                                                 NaN   
6                                                 NaN   
7                                                 NaN   
8                                                 NaN   
9                                                 NaN   
10                                                NaN   
11                                                NaN   
12                                                NaN   
13                                                NaN   
14                                                NaN   
15                                                NaN   
16                                                NaN   
17                                                NaN   
18                                                NaN   
19                                                NaN   
20                                                NaN   
21                                                NaN   
22                                                NaN   
23                                                NaN   
24                                                NaN   
25                                                NaN   
26                                                NaN   
27                                                NaN   
28                                                NaN   
29                                                NaN   
30                                                NaN   
31                                                NaN   
32                                                NaN   
33                                                NaN   
34                                                NaN   
35                                                NaN   
36                                                NaN   
37                                                NaN   
38                                                NaN   

                                           Unnamed: 1 Unnamed: 2 Unnamed: 3  \
0                                                 NaN        NaN        NaN   
1          Tresiba® demand volume overview (YTD May*)        NaN        NaN   
2                                                 NaN        NaN        NaN   
3                        Tresiba® demand vol.\n(mXPO)   Realized     RE19/1   
4                                                 NaN        NaN        NaN   
5                                        Total volume    2720.97    2713.68   
6                                                 NaN        NaN              
7                                                 NaN        NaN              
8                        Total share of basal segment   0.135579   0.135247   
9                                                 NaN        NaN              
10                                                NaN        NaN              
11                                  Commercial volume    1557.52    1558.09   
12                                                NaN        NaN              
13                                                NaN        NaN              
14                                      Part D volume    1006.14    999.547   
15                                                NaN        NaN        NaN   
16                                                NaN        NaN        NaN   
17                                                NaN        NaN        NaN   
18                                                NaN        NaN        NaN   
19    *March 2019 represents the most current mont...        NaN        NaN   
20                                                NaN        NaN        NaN   
21  Retail: Tresiba® sou

### 6B. Key brand performance

In [196]:
may19_6B = pd.read_excel("QAP May 2019 NNI Performance Business Update v2.xlsm", sheet_name="6B. Key brand performance")
may19_6B

Unnamed: 0                                         Unnamed: 1  \
0          NaN                              Key brand performance   
1          NaN   Ex-factory: May 2019   /   In-market: April 2019   
2          NaN                                                NaN   
3          NaN                                                YTD   
4          NaN                                                NaN   
5          NaN                                 Diabetes & Obesity   
6          NaN                                           Victoza®   
7          NaN                                           Ozempic®   
8          NaN                                           Tresiba®   
9          NaN                                           Saxenda®   
10         NaN                                          Xultophy®   
11         NaN                                           Levemir®   
12         NaN                                           NovoLog®   
13         NaN                                             Fiasp®   
14         NaN                                       NovoLog® Mix   
15         NaN                                           Biopharm   
16         NaN                                         NovoSeven®   
17         NaN                                         Novoeight®   
18         NaN                                       Norditropin®   
19         NaN                                                NaN   
20         NaN                                                MTD   
21         NaN                                                NaN   
22         NaN                                 Diabetes & Obesity   
23         NaN                                           Victoza®   
24         NaN                                           Ozempic®   
25         NaN                                           Tresiba®   
26         NaN                                           Saxenda®   
27         NaN                                          Xultophy®   
28         NaN                                           Levemir®   
29         NaN                                           NovoLog®   
30         NaN                                             Fiasp®   
31         NaN                                       NovoLog® Mix   
32         NaN                                           Biopharm   
33         NaN                                         NovoSeven®   
34         NaN                                         Novoeight®   
35         NaN                                       Norditropin®   
36         NaN                                                NaN   
37         NaN                                                NaN   
38         NaN                                                NaN   
39         NaN                                                NaN   
40         NaN                                                NaN   
41         NaN           Source: SAP (Ex-Factory), IQVIA-Xponent.   
42         NaN  Ex-Factory Volume: MU (insulin); SU (Victoza®,...   
43         NaN                                                NaN   
44         NaN  Commercial include Managed Care+Mail; Part D i...   
45         NaN  Index Color: red < 99.0 ; orange 99.0 to 99.9 ...   

          Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
0                NaN        NaN        NaN        NaN        NaN   
1                NaN        NaN        NaN        NaN        NaN   
2                NaN        NaN        NaN        NaN        NaN   
3   Net Sales (mUSD)        NaN        NaN        NaN        NaN   
4                REA     RE19/1      Index   RE18/418      Index   
5            2392.47    2442.15    97.9656          0        NaN   
6              756.7    773.712    97.8013          0        NaN   
7            280.494    270.602    103.655          0        NaN   
8            264.899    272.884    97.0742          0        NaN   
9            138.423      141.9    97.5499          0        NaN   
10          

### 6C. In market perf. - total

In [197]:
may19_6C = pd.read_excel("QAP May 2019 NNI Performance Business Update v2.xlsm", sheet_name="6C. In market perf. - total")
may19_6C

Unnamed: 0                                Unnamed: 1   Unnamed: 2  \
0          NaN                                       NaN          NaN   
1          NaN      In market performance\nIQVIA Xponent          NaN   
2          NaN        latest available month: March 2019          NaN   
3          NaN                                       NaN          NaN   
4          NaN                                       NaN          REA   
5          NaN    Market Segment Volume and Growth (XPO)   YTD Apr'19   
6          NaN                    Insulin total (XPO MU)      37600.9   
7          NaN                    Basal insulin (XPO MU)      20069.2   
8          NaN              Fast Acting insulin (XPO MU)      14338.5   
9          NaN                   Premix insulin (XPO MU)      3193.08   
10         NaN                                       NaN          NaN   
11         NaN              NG+MI Insulin total (XPO MU)      33053.7   
12         NaN              NG+MI Basal insulin (XPO MU)      18996.4   
13         NaN        NG+MI Fast Acting insulin (XPO MU)      12377.5   
14         NaN             NG+MI Premix insulin (XPO MU)      1679.87   
15         NaN                                       NaN          NaN   
16         NaN                      Modern NIAD (XPO SU)  1.93158e+07   
17         NaN                            DPP-4 (XPO SU)  8.71479e+06   
18         NaN                            GLP-1 (XPO SU)  5.19776e+06   
19         NaN                           SGLT-2 (XPO SU)   5.2567e+06   
20         NaN                   SGLT-2 + DPP-4 (XPO SU)       146513   
21         NaN                                       NaN          NaN   
22         NaN                                       NaN          REA   
23         NaN       Novo Nordisk US Market Shares (XPO)   YTD Apr'19   
24         NaN                   Value Share of Diabetes     0.289611   
25         NaN              Value Share of total Insulin     0.401012   
26         NaN              Value Share of NG+MI Insulin     0.412231   
27         NaN                                       NaN          NaN   
28         NaN             Volume share of total insulin     0.409298   
29         NaN             Volume share of NG+MI insulin        0.412   
..         ...                                       ...          ...   
43         NaN                                     GLP-1     0.269094   
44         NaN                    GLP-1 - QD/BID segment    0.0945855   
45         NaN                        GLP-1 - QW segment     0.174509   
46         NaN                                  Victoza®     0.091039   
47         NaN                                  Ozempic®    0.0275418   
48         NaN                    lixisenatide (Adlyxin)  3.22876e-05   
49         NaN                    exenatide BID (Byetta)   0.00351426   
50         NaN                   exenatide ER (Bydureon)    0.0245413   
51         NaN                   dulaglutide (Trulicity)     0.122357   
52         NaN                     albiglutide (Tanzeum)  6.88654e-05   
53         NaN                                       NaN          NaN   
54         NaN                                    SGLT-2     0.272146   
55         NaN     canagliflozin (Invokana, Invokanamet)          NaN   
56         NaN           dapagliflozin (Farxiga, Xigduo)          NaN   
57         NaN      emplagliflozin (Jardiance, Synjardy)          NaN   
58         NaN                                       NaN          NaN   
59         NaN                            SGLT-2 + DPP-4   0.00758513   
60         NaN    empagliflozin + linagliptin (Glyxambi)          NaN   
61         NaN                                       NaN          NaN   
62         NaN                                     DPP-4     0.451175   
63         NaN  sitagliptin (Januvia, Janumet, Juvisync)          NaN   
64         NaN         saxagliptin (Onglyza, Kombiglyze)          NaN   
65         NaN        linagliptin (Trajenta, Jentadueto)      

### 6D. Monthly in market - total

In [198]:
may19_6D = pd.read_excel("QAP May 2019 NNI Performance Business Update v2.xlsm", sheet_name="6D. Monthly in market - total")
may19_6D

Unnamed: 0                                        Unnamed: 1  \
0           NaN                                               NaN   
1           NaN  Monthly in market TRx performance\nIQVIA Xponent   
2           NaN                latest available month: March 2019   
3           NaN                                               NaN   
4           NaN                                               NaN   
5           NaN                                               NaN   
6           NaN                                               NaN   
7           NaN                                          Tresiba®   
8           NaN                                               NaN   
9           NaN                                               NaN   
10          NaN                                               NaN   
11          NaN                                               NaN   
12          NaN                                               NaN   
13          NaN                                               NaN   
14          NaN                                               NaN   
15          NaN                                               NaN   
16          NaN                                               NaN   
17          NaN                                               NaN   
18          NaN                                               NaN   
19          NaN                                          Levemir®   
20          NaN                                               NaN   
21          NaN                                               NaN   
22          NaN                                               NaN   
23          NaN                                               NaN   
24          NaN                                               NaN   
25          NaN                                               NaN   
26          NaN                                               NaN   
27          NaN                                               NaN   
28          NaN                                               NaN   
29          NaN                                               NaN   
..          ...                                               ...   
72          NaN                                               NaN   
73          NaN                                               NaN   
74          NaN                                               NaN   
75          NaN                                               NaN   
76          NaN                                               NaN   
77          NaN                                               NaN   
78          NaN                                               NaN   
79          NaN                                          Saxenda®   
80          NaN                                               NaN   
81          NaN                                               NaN   
82          NaN                                               NaN   
83          NaN                                               NaN   
84          NaN                                               NaN   
85          NaN                                               NaN   
86          NaN                                               NaN   
87          NaN                                               NaN   
88          NaN                                               NaN   
89          NaN                                               NaN   
90          NaN                                               NaN   
91          NaN                                               NaN   
92          NaN                                               NaN   
93          NaN                                               NaN   
94          NaN                                               NaN   
95          NaN                                               NaN   
96          NaN                                               NaN   
97          NaN                                               NaN   
98 

# update function to update 3D, 6B, 6C, 6D

In [199]:
def update(*args):
    may19_3D.iloc[5,3] = "QAPPPPPP"
    may19_3D.iloc[8,3] = tr[1]
    may19_3D.iloc[11,3] = tr[2]
    may19_3D.iloc[14,3] = tr[3]
    ########
    may19_6B.iloc[8,13] = tr[0]
    may19_6B.iloc[8,18] = tr[1]
    may19_6B.iloc[8,23] = tr[2]
    may19_6B.iloc[8,28] = tr[3]
    ########
    may19_6C.iloc[30,12] = tr[1]
    may19_6C.iloc[30,14] = tr[6]
    ########
    may19_6D.iloc[14,20] = tr[1]
    may19_6D.iloc[14,21] = tr[6]

## replaced a cell in 3D with QAPPPPPP for testing

In [200]:
update(may19_3D, may19_6B, may19_6C, may19_6D)

In [201]:
may19_3D

Unnamed: 0  \
0   Tresiba® continues strong volume and share per...   
1                                                 NaN   
2                                                 NaN   
3                                                 NaN   
4                                                 NaN   
5                                                 NaN   
6                                                 NaN   
7                                                 NaN   
8                                                 NaN   
9                                                 NaN   
10                                                NaN   
11                                                NaN   
12                                                NaN   
13                                                NaN   
14                                                NaN   
15                                                NaN   
16                                                NaN   
17                                                NaN   
18                                                NaN   
19                                                NaN   
20                                                NaN   
21                                                NaN   
22                                                NaN   
23                                                NaN   
24                                                NaN   
25                                                NaN   
26                                                NaN   
27                                                NaN   
28                                                NaN   
29                                                NaN   
30                                                NaN   
31                                                NaN   
32                                                NaN   
33                                                NaN   
34                                                NaN   
35                                                NaN   
36                                                NaN   
37                                                NaN   
38                                                NaN   

                                           Unnamed: 1 Unnamed: 2 Unnamed: 3  \
0                                                 NaN        NaN        NaN   
1          Tresiba® demand volume overview (YTD May*)        NaN        NaN   
2                                                 NaN        NaN        NaN   
3                        Tresiba® demand vol.\n(mXPO)   Realized     RE19/1   
4                                                 NaN        NaN        NaN   
5                                        Total volume    2720.97   QAPPPPPP   
6                                                 NaN        NaN              
7                                                 NaN        NaN              
8                        Total share of basal segment   0.135579   0.135247   
9                                                 NaN        NaN              
10                                                NaN        NaN              
11                                  Commercial volume    1557.52    1558.09   
12                                                NaN        NaN              
13                                                NaN        NaN              
14                                      Part D volume    1006.14    999.547   
15                                                NaN        NaN        NaN   
16                                                NaN        NaN        NaN   
17                                                NaN        NaN        NaN   
18                                                NaN        NaN        NaN   
19    *March 2019 represents the most current mont...        NaN        NaN   
20                                                NaN        NaN        NaN   
21  Retail: Tresiba® sou

## under RE19/1 and NaN in 4th column, you'll see QAPPPPPP (was 2714)

In [202]:
may19_6B

Unnamed: 0                                         Unnamed: 1  \
0          NaN                              Key brand performance   
1          NaN   Ex-factory: May 2019   /   In-market: April 2019   
2          NaN                                                NaN   
3          NaN                                                YTD   
4          NaN                                                NaN   
5          NaN                                 Diabetes & Obesity   
6          NaN                                           Victoza®   
7          NaN                                           Ozempic®   
8          NaN                                           Tresiba®   
9          NaN                                           Saxenda®   
10         NaN                                          Xultophy®   
11         NaN                                           Levemir®   
12         NaN                                           NovoLog®   
13         NaN                                             Fiasp®   
14         NaN                                       NovoLog® Mix   
15         NaN                                           Biopharm   
16         NaN                                         NovoSeven®   
17         NaN                                         Novoeight®   
18         NaN                                       Norditropin®   
19         NaN                                                NaN   
20         NaN                                                MTD   
21         NaN                                                NaN   
22         NaN                                 Diabetes & Obesity   
23         NaN                                           Victoza®   
24         NaN                                           Ozempic®   
25         NaN                                           Tresiba®   
26         NaN                                           Saxenda®   
27         NaN                                          Xultophy®   
28         NaN                                           Levemir®   
29         NaN                                           NovoLog®   
30         NaN                                             Fiasp®   
31         NaN                                       NovoLog® Mix   
32         NaN                                           Biopharm   
33         NaN                                         NovoSeven®   
34         NaN                                         Novoeight®   
35         NaN                                       Norditropin®   
36         NaN                                                NaN   
37         NaN                                                NaN   
38         NaN                                                NaN   
39         NaN                                                NaN   
40         NaN                                                NaN   
41         NaN           Source: SAP (Ex-Factory), IQVIA-Xponent.   
42         NaN  Ex-Factory Volume: MU (insulin); SU (Victoza®,...   
43         NaN                                                NaN   
44         NaN  Commercial include Managed Care+Mail; Part D i...   
45         NaN  Index Color: red < 99.0 ; orange 99.0 to 99.9 ...   

          Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
0                NaN        NaN        NaN        NaN        NaN   
1                NaN        NaN        NaN        NaN        NaN   
2                NaN        NaN        NaN        NaN        NaN   
3   Net Sales (mUSD)        NaN        NaN        NaN        NaN   
4                REA     RE19/1      Index   RE18/418      Index   
5            2392.47    2442.15    97.9656          0        NaN   
6              756.7    773.712    97.8013          0        NaN   
7            280.494    270.602    103.655          0        NaN   
8            264.899    272.884    97.0742          0        NaN   
9            138.423      141.9    97.5499          0        NaN   
10          

In [215]:
pd.DataFrame([may19_6C.iloc[5,:], may19_6C.iloc[30,:]])

Unnamed: 0                              Unnamed: 1  Unnamed: 2  \
5          NaN  Market Segment Volume and Growth (XPO)  YTD Apr'19   
30         NaN          Tresiba® volume share of basal    0.135579   

    Unnamed: 3      Unnamed: 4               Unnamed: 5  Unnamed: 6  \
5   MAT Apr'19  y/y MAT Growth  y/y MAT Share of Growth         NaN   
30    0.126661        0.138505                      NaN         NaN   

    Unnamed: 7  Unnamed: 8      Unnamed: 9          Unnamed: 10  Unnamed: 11  \
5   YTD Mar'17  MAT Mar'17  y/y MAT Growth  Ful year MAT growth          NaN   
30         NaN         NaN             NaN            0.0208291          NaN   

   Unnamed: 12 Unnamed: 13     Unnamed: 14            Unnamed: 15  \
5   YTD Mar'19  MAT Mar'19  y/y MAT Growth  Full year\nMAT growth   
30    0.135247    0.127509        0.140641               0.141966   

    Unnamed: 16  Unnamed: 17  Unnamed: 18  
5           NaN          NaN          NaN  
30          NaN          NaN          NaN

In [204]:
may19_6D

Unnamed: 0                                        Unnamed: 1  \
0           NaN                                               NaN   
1           NaN  Monthly in market TRx performance\nIQVIA Xponent   
2           NaN                latest available month: March 2019   
3           NaN                                               NaN   
4           NaN                                               NaN   
5           NaN                                               NaN   
6           NaN                                               NaN   
7           NaN                                          Tresiba®   
8           NaN                                               NaN   
9           NaN                                               NaN   
10          NaN                                               NaN   
11          NaN                                               NaN   
12          NaN                                               NaN   
13          NaN                                               NaN   
14          NaN                                               NaN   
15          NaN                                               NaN   
16          NaN                                               NaN   
17          NaN                                               NaN   
18          NaN                                               NaN   
19          NaN                                          Levemir®   
20          NaN                                               NaN   
21          NaN                                               NaN   
22          NaN                                               NaN   
23          NaN                                               NaN   
24          NaN                                               NaN   
25          NaN                                               NaN   
26          NaN                                               NaN   
27          NaN                                               NaN   
28          NaN                                               NaN   
29          NaN                                               NaN   
..          ...                                               ...   
72          NaN                                               NaN   
73          NaN                                               NaN   
74          NaN                                               NaN   
75          NaN                                               NaN   
76          NaN                                               NaN   
77          NaN                                               NaN   
78          NaN                                               NaN   
79          NaN                                          Saxenda®   
80          NaN                                               NaN   
81          NaN                                               NaN   
82          NaN                                               NaN   
83          NaN                                               NaN   
84          NaN                                               NaN   
85          NaN                                               NaN   
86          NaN                                               NaN   
87          NaN                                               NaN   
88          NaN                                               NaN   
89          NaN                                               NaN   
90          NaN                                               NaN   
91          NaN                                               NaN   
92          NaN                                               NaN   
93          NaN                                               NaN   
94          NaN                                               NaN   
95          NaN                                               NaN   
96          NaN                                               NaN   
97          NaN                                               NaN   
98 

## 3D, 6B, 6C, 6D updated



## library to work with existing workbook, edit, and save
### creating empty workbook to try and add 3D.basal sheet, 6B, 6C, 6D

In [205]:
from openpyxl import Workbook #create workbook
from openpyxl import load_workbook #existing workbook

In [247]:
vl = load_workbook(filename = "TEST.xlsx")

In [248]:
vl.create_sheet("3D. Basal in-market")
vl.create_sheet("6B. Key brand performance")
vl.create_sheet("6C. In market perf. - total")
vl.create_sheet("6D. Monthly in market - total")

<Worksheet "6D. Monthly in market - total">

In [249]:
vl.save("TEST.xlsx")

In [250]:
def append_df_to_excel(filename, df, sheet_name, startrow=None,
                       truncate_sheet=False, 
                       **to_excel_kwargs):

    # ignore [engine] parameter if it was passed
    if 'engine' in to_excel_kwargs:
        to_excel_kwargs.pop('engine')

    writer = pd.ExcelWriter(filename, engine='openpyxl')

    # Python 2.x: define [FileNotFoundError] exception if it doesn't exist 
    try:
        FileNotFoundError
    except NameError:
        FileNotFoundError = IOError

    try:
        # try to open an existing workbook
        writer.book = load_workbook(filename)

        # get the last row in the existing Excel sheet
        # if it was not specified explicitly
        if startrow is None and sheet_name in writer.book.sheetnames:
            startrow = writer.book[sheet_name].max_row

        # truncate sheet
        if truncate_sheet and sheet_name in writer.book.sheetnames:
            # index of [sheet_name] sheet
            idx = writer.book.sheetnames.index(sheet_name)
            # remove [sheet_name]
            writer.book.remove(writer.book.worksheets[idx])
            # create an empty sheet [sheet_name] using old index
            writer.book.create_sheet(sheet_name, idx)

        # copy existing sheets
        writer.sheets = {ws.title:ws for ws in writer.book.worksheets}
    except FileNotFoundError:
        # file does not exist yet, we will create it
        pass

    if startrow is None:
        startrow = 0

    # write out the new sheet
    df.to_excel(writer, sheet_name, startrow=startrow, **to_excel_kwargs)

    # save the workbook
    writer.save()

In [251]:
append_df_to_excel(filename="TEST.xlsx", df=may19_3D, sheet_name='3D. Basal in-market')
append_df_to_excel(filename="TEST.xlsx", df=may19_6B, sheet_name='6B. Key brand performance')
append_df_to_excel(filename="TEST.xlsx", df=may19_6C, sheet_name='6C. In market perf. - total')
append_df_to_excel(filename="TEST.xlsx", df=may19_6D, sheet_name='6D. Monthly in market - total')

# WORKED! :)
## can change color and texture too (saw somewhere)